go/vt/topo/test/trylock.go

```diff
@@ -0,0 +1,182 @@
+		if err != nil {
+			require.Fail(t, "ListDir failed", err.Error())
+		}
+		for _, e := range entries {
+			if e.Ephemeral {
+				t.Logf("skipping ephemeral node %v in %v", e, p)
+			} else {
+				require.Fail(t, "non-ephemeral DirEntry")
+			}
+		}
+	}
+
+	// test we can't take the lock again
+	fastCtx, cancel := context.WithTimeout(ctx, 10*time.Second)
+	if _, err := conn.TryLock(fastCtx, keyspacePath, "again"); !topo.IsErrType(err, topo.NodeExists) {
+		require.Fail(t, "TryLock failed", err.Error())
+	}
+	cancel()
+
+	// test we can interrupt taking the lock
+	interruptCtx, cancel := context.WithCancel(ctx)
+	go func() {
+		time.Sleep(1 * time.Second)
+		cancel()
+	}()
+	if _, err := conn.TryLock(interruptCtx, keyspacePath, "interrupted"); !topo.IsErrType(err, topo.NodeExists) {
+		require.Fail(t, "TryLock failed", err.Error())
+	}
+
+	if err := lockDescriptor.Check(ctx); err != nil {
+		t.Errorf("Check(): %v", err)
+	}
+
+	if err := lockDescriptor.Unlock(ctx); err != nil {
+		require.Fail(t, "Unlock failed", err.Error())
+	}
+
+	// test we can't unlock again
+	if err := lockDescriptor.Unlock(ctx); err == nil {
+		require.Fail(t, "Unlock failed", err.Error())
+	}
+}
+
+// checkTryLockMissing makes sure we can't lock a non-existing directory.
+func checkTryLockMissing(ctx context.Context, t *testing.T, conn topo.Conn) {
+	keyspacePath := path.Join(topo.KeyspacesPath, "test_keyspace_666")
+	if _, err := conn.TryLock(ctx, keyspacePath, "missing"); err == nil {
+		require.Fail(t, "TryLock(test_keyspace_666) worked for non-existing keyspace")
+	}
+}
+
+// unlike 'checkLockUnblocks', checkTryLockUnblocks will not block on other client but instead
+// keep retrying until it gets the lock.
+func checkTryLockUnblocks(ctx context.Context, t *testing.T, conn topo.Conn) {
+	keyspacePath := path.Join(topo.KeyspacesPath, "test_keyspace")
+	unblock := make(chan struct{})
+	finished := make(chan struct{})
+
+	// TryLock will keep getting NodeExists until lockDescriptor2 unlock itself.
+	// It will not wait but immediately return with NodeExists error.
+	go func() {
+		<-unblock
+		waitUntil := time.Now().Add(10 * time.Second)
+		for time.Now().Before(waitUntil) {
+			lockDescriptor, err := conn.TryLock(ctx, keyspacePath, "unblocks")
+			if err != nil {
+				if !topo.IsErrType(err, topo.NodeExists) {
+					require.Fail(t, "expected node exists during trylock", err.Error())
+				}
+				time.Sleep(1 * time.Second)
+			} else {
+				if err = lockDescriptor.Unlock(ctx); err != nil {
+					require.Fail(t, "Unlock(test_keyspace) failed", err.Error())
+				}
+				close(finished)
+				break
+			}
+		}
+	}()
+
+	// Lock the keyspace.
+	lockDescriptor2, err := conn.TryLock(ctx, keyspacePath, "")
+	if err != nil {
+		require.Fail(t, "Lock(test_keyspace) failed", err.Error())
+	}
+
+	// unblock the go routine so it starts waiting
+	close(unblock)
+
+	// sleep for a while so we're sure the go routine is blocking
+	time.Sleep(5 * time.Second)
```

Comment:
Same comment here. I don't see the point of this. What are we actually waiting for?